DisallowMultipleAttributesPerLineSniff

michnovka opened the pull request.

michnovka commented:
@kukulich I dont think I can really improve the indentation efficiently. There is just not "correct" way of doing it, this should be enforced using other sniffs. What further complicates things is that we register the sniff on `T_ATTRIBUTE` so when I fix one attribute, I have now inconsistent actual fixed file with tokens from phpcs, as those point to the original parsed file.

I do not have more time to work on this, I am happy with the sniff as is. If you have some more suggestions, let me know.

kukulich commented:
@michnovka I really don't understand why the sniff is done differently than suggested in https://github.com/slevomat/coding-standard/issues/1442#issuecomment-1276112153

michnovka commented:
when implementing this I was facing some difficulties. Manily due to fixer and the fact this sniff targets `T_ATTRIBUTE`. By the time I fixed one attribute (e.g. add new line after it) it would modify the file for the 2nd one following and suddenly the tokens from phpcsfile were outdated (not corresponding to the fix-in-progress file). I couldnt make it work, so I asked myself the question - whats is the main purpose of this sniff? And that is to not allow having multiple attributes on one line. So thats what I implemented.

I am happy with renaming the sniff to `RequireAttributeOnDedicatedLine` but without targetting directly `AttributeHelper::attributableTokens` instead of T_ATTRIBUTEs I have hard time making it work as in the original comment.

kukulich commented:
> I am happy with renaming the sniff to RequireAttributeOnDedicatedLine

I'm not. That's not what I suggested.

michnovka commented:
Another issue is mixing of phpdoc and attributes. How to handle spacing there in regards to `linesCountBetweenAttributes`?

This sniff makes sure that attributes are always the only thing on a line. There is an exception for single in-line attributes

michnovka commented:
@kukulich I have edited this PR based on https://github.com/slevomat/coding-standard/issues/1442#issuecomment-1279341581 Please review

kukulich commented on SlevomatCodingStandard/Sniffs/Attributes/DisallowMultipleAttributesPerLineSniff.php:
```diff
@@ -0,0 +1,89 @@
+			$tokens[$attributeCloserPointer]['line'] === $tokens[$nextNonWhitespaceAfterAttributeCloserPointer]['line']
+		) {
+			$fixNotAloneOnLine = $phpcsFile->addFixableError(
+				'Attribute is not alone on line',
```
`Multiple attributes per line are disallowed.`

michnovka commented on tests/Sniffs/Attributes/data/disallowedMultipleAttributesPerLineErrors.fixed.php:
```diff
@@ -0,0 +1,101 @@
+#[Attribute3(option: PDO::class, option2: true, option3: 'False')]
+#[Attribute4()]
+#[Attribute5]
+// comment
```
@kukulich is there some helper function similar to `TokenHelper::findFirstNonWhitespaceOnLine()` that would look for non-whitespace AND non-comment?

kukulich replied:
You can just check if there's another `T_ATTRIBUTE` on the same line before current attribute.

michnovka replied:
no, I cannot. imagine:

```php
#[Attribute1]
#[Attribute2] /** whatever */ public int $x;
```

kukulich replied:
I don't see any problem.

michnovka replied:
it should be fixed into
```php
#[Attribute1]
#[Attribute2] /** whatever */
public int $x;
```

the point of this sniff is to have on one line ONLY one attribute AND whitespaces OR comments. But OK, we can make this sniff ONLY about 2 attributes on one line and make another sniff which disallows having attribute and attributed token on the same line. WDYT?

kukulich replied:
> make another sniff which disallows having attribute and attributed token on the same line. WDYT?

That's the sniff that should solve number of lines between last attribute and owner (or comment).

michnovka replied:
> That's the sniff that should solve number of lines between last attribute and owner (or comment).
How would you call that sniff?

kukulich replied:
https://github.com/slevomat/coding-standard/commit/3b165784080535b827f6fa5c311d23d64fe38b7f

kukulich commented:
@michnovka Thanks! I've finished the sniff by myself.
